velox/docs/develop/tracing.rst

```diff
@@ -0,0 +1,152 @@
+=======
+Tracing
+=======
+
+
+Background
+----------
+
+The query trace tool helps analyze and debug query performance and correctness
+issues. It helps prevent interference from the test noises in a production
+environment (such as storage, network etc) by allowing replay of a part of the
+query plan and data set in an isolated environment such as a local machine.
+This is much more efficient for query performance debugging as we don't have to
+replay the whole query in a production environment.
+
+How Tracing Tool Works
+----------------------
+
+The tracing process consists of two phases: tracing and replaying.
+
+**Tracing Phase**
+
+- When the query starts, the task records the metadata including query plan fragment,
+  query configuration, and connector properties.
+- During the query running, each traced operator records the input vectors and saves
+  in the specified storage location.
+- The metadata are serialized using json format and operator data inputs are serialized
+  using Presto serializer.
+
+**Replaying Phase**
+
+- Read and deserialize the recorded query plan in tracing phase, extract the target plan
+  node, and assemble a plan fragment with a customized source and sink nodes. The source
+  node reads the input from the serialized operator inputs on storage and sink operator
+  prints or logs out the execution stats.
+- Build a query/task with the assembled plan fragment in step 1.
+- Apply the recorded query configuration and connector properties to replay the query/task
+  with the same input and configuration setup as in production.
+
+**NOTE**: the presto serialization might lose the input vector encoding such as lazy vector
```

Comment:
I feel this tool is most useful for replying non-leaf fragment of the query. For leaf fragments it is much simpler to replay using original splits. 

Let's restrict this tool to non-leaf fragments and not worry about connectors.

For leaf fragments, we can still record the plan fragment and splits, but let's not record data.

For non-leaf fragments, let's record plan fragment and write exchange data to files on disk. 

I think it is important to have limits on the amount of data written to disk and a filter on task ID that records the data from the start. Otherwise, enabling tracing for a single query in a multi-tenant prod cluster can bring the cluster down.

CC: @xiaoxmeng

Reply:
@mbasmanova yes, we shall just record splits for replay on leaf-fragment but no data. @duanmeng let's add total input data size limit check as what we do for spilling to avoid writing too much data in production even we store the tracing data on remote storage.

@mbasmanova we probably doesn't need to have limit on task id in Velox. It could be configured by Prestissimo worker when used together. Like Prestissimo extract shard id from task id and only configure tasks for tracing which has the shard id match. WDYT? Thanks!

Reply:
>For leaf fragments, we can still record the plan fragment and splits, but let's not record data.

Yes, that's the plan that only records splits when we add the support of `TableScan`, in which the QueryDataWriter would not be created.

>I think it is important to have limits on the amount of data written to disk and a filter on task ID that records the data from the start. 

Got it, I've added two query configs, `max_trace_bytes` and `query_trace_task_reg_exp`.

@mbasmanova @xiaoxmeng